src/Action.cpp

```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
     }
 }
 
-void BaseAction::complete()
-{
+void BaseAction::complete() {
     status = COMPLETED;
 }
 
-void BaseAction::error(const std::string& errorMsg)
-{
+void BaseAction::error(const std::string& errorMsg) {
     this->errorMsg = errorMsg;
     status = ERROR;
 }
 
-std::string BaseAction::getErrorMsg() const
-{
+std::string BaseAction::getErrorMsg() const {
     return errorMsg;
 }
 
-CreateUser::CreateUser(const std::string& name,
-                         const std::string& recommendationType):
-                         name(name), recommendationType(recommendationType) 
-{}
+CreateUser::CreateUser(const std::string& name, const std::string& recommendationAlg): BaseAction(),
+                                                                                        _userName(name),
+                                                                    _recommendationAlg(recommendationAlg) {}
 
-void CreateUser::act(Session& sess)
-{
-    const std::unordered_map<std::string, User*>& users = sess.getUsers();
-    if(users.find(name) != users.end())
-    {
+void CreateUser::act(Session& sess) {
+    if(sess.getUsers().count(_userName)) {
+        error(USER_EXISTS_ERR);
+        return;
+    }
+    if (_recommendationAlg == "len") {
+        sess.addToUserMap(new LengthRecommenderUser(_userName));
+    }
+    else if (_recommendationAlg == "rer") {
+        sess.addToUserMap(new RerunRecommenderUser(_userName));
+    }
+    else if (_recommendationAlg == "gen") {
+        sess.addToUserMap(new GenreRecommenderUser(_userName));
+    }
+    else {
+        error(INVALID_ALG_ERR);
+        return;
+    }
+    complete();
+}
+std::string CreateUser::toString() const {
+    return ("creating new User - " + _userName); //TODO complete due to forum answer
+}
+
+ChangeActiveUser::ChangeActiveUser(const std::string& name): BaseAction(), _newUser(name) {}
 
+void ChangeActiveUser::act(Session &sess) {
+    const std::unordered_map<std::string,User*>& usersMap = sess.getUsers();
+    if(!usersMap.count(_newUser)) {
+        error(USER_DOESNT_EXISTS_ERR);
+    }
+    else {
+        sess.setActiveUser(*(usersMap.at(_newUser)));
+        complete();
```

Comment:
using find instead of count here would mean you could iterate the map only once

Reply:
don't understand. count is "Logarithmic in size" - from cpp reference.

Reply:
Complexity won't change. But finding the same element twice (once with count and once with at) seems quicker, albeit not **much** quicker